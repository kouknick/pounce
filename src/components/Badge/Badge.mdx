A Badge should be used to render statuses

A badge can be `solid`:

```typescript jsx
<SimpleGrid spacing={3} columns={3}>
  <Badge color="red-200">INFO</Badge>
  <Badge color="teal-300">LOW</Badge>
  <Badge color="blue-600">WHATEVER</Badge>
</SimpleGrid>
```

A badge can be `outline`:

```typescript jsx
<SimpleGrid spacing={3} columns={3}>
  <Badge color="red-200" variant="outline">
    INFO
  </Badge>
  <Badge color="teal-300" variant="outline">
    LOW
  </Badge>
  <Badge color="blue-600" variant="outline">
    WHATEVER
  </Badge>
</SimpleGrid>
```
