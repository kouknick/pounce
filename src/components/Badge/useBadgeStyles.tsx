import { BadgeProps } from './Badge';
import { FlexProps } from '../Flex';

type UseBadgeStylesProps = Pick<BadgeProps, 'variant' | 'color'>;
type UseBadgeStylesPayload = Partial<FlexProps>;

const useBadgeStyles = ({ variant, color }: UseBadgeStylesProps): UseBadgeStylesPayload => {
  switch (variant) {
    case 'outline':
      return {
        border: '1px solid',
        borderColor: color,
        bg: 'transparent' as const,
        color: 'white',
      };
    case 'solid':
    default:
      return { bg: color, color: 'gray-50' };
  }
};

export default useBadgeStyles;
