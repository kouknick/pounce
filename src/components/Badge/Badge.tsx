import React from 'react';
import Flex from '../Flex';
import { Theme } from '../../theme';
import useBadgeStyles from './useBadgeStyles';

export interface BadgeProps {
  /** The color theme of the badge */
  color: keyof Theme['colors'];

  /** The style of the badge */
  variant?: 'solid' | 'outline';
}

/** A badge is simply a visual label to accompany & characterize a certain text*/
const Badge = React.forwardRef<HTMLElement, BadgeProps>(function Badge(
  { color, variant = 'solid', children, ...rest },
  ref
) {
  const variantStyles = useBadgeStyles({ variant, color });

  return (
    <Flex
      ref={ref}
      cursor="default"
      width="fit-content"
      minWidth="85px"
      textAlign="center"
      fontWeight="bold"
      borderRadius="pill"
      align="center"
      justify="center"
      fontSize="small"
      py={1}
      px={4}
      {...variantStyles}
      {...rest}
    >
      {children}
    </Flex>
  );
});

export default Badge;
