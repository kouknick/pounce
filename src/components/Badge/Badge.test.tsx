import React from 'react';
import { render, act, waitFor } from 'test-utils';
import { axe, AxeResults } from 'jest-axe';

import Badge from './Badge';

it('renders', () => {
  const { container } = render(
    <>
      <Badge color="red-300">INFO</Badge>
      <Badge color="teal-500">LOW</Badge>
      <Badge color="blue-400">WHATEVER</Badge>
    </>
  );
  expect(container).toMatchSnapshot();
});

it('aligns with a11y', async () => {
  const { container } = render(
    <>
      <Badge color="red-300">INFO</Badge>
      <Badge color="teal-500">LOW</Badge>
      <Badge color="blue-400">WHATEVER</Badge>
    </>
  );
  let results: AxeResults;
  await act(async () => {
    results = await axe(container);
  });
  await waitFor(() => {
    expect(results).toHaveNoViolations();
  });
});
