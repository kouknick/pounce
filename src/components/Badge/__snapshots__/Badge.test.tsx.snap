// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.pounce-0 {
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 32px;
  padding-right: 32px;
  background-color: red-300;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-fit-content;
  width: -moz-content;
  width: fit-content;
  min-width: 85px;
  border-radius: pill;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  font-weight: bold;
  font-size: small;
  cursor: default;
}

.pounce-2 {
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 32px;
  padding-right: 32px;
  background-color: teal-500;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-fit-content;
  width: -moz-content;
  width: fit-content;
  min-width: 85px;
  border-radius: pill;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  font-weight: bold;
  font-size: small;
  cursor: default;
}

.pounce-4 {
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 32px;
  padding-right: 32px;
  background-color: blue-400;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-fit-content;
  width: -moz-content;
  width: fit-content;
  min-width: 85px;
  border-radius: pill;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  font-weight: bold;
  font-size: small;
  cursor: default;
}

<div>
  <div
    aria-atomic="true"
    class="pounce-0 pounce-1"
    role="status"
  >
    INFO
  </div>
  <div
    aria-atomic="true"
    class="pounce-2 pounce-1"
    role="status"
  >
    LOW
  </div>
  <div
    aria-atomic="true"
    class="pounce-4 pounce-1"
    role="status"
  >
    WHATEVER
  </div>
</div>
`;
